File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Set<List<Integer>> ans = new HashSet<>();
        Arrays.sort(nums);

        for(int i=0;i<nums.length;i++){
            int sum = nums[i];
            int start = i+1;
            int end = nums.length-1;

            while(start < end){
                if((sum + nums[start] + nums[end]) == 0){
                    ArrayList<Integer> al = new ArrayList<>();
                    al.add(sum);
                    al.add(nums[start]);
                    al.add(nums[end]);
                    ans.add(al);
                    start++;
                    end--;
                }
                else if((sum + nums[start] + nums[end]) < 0) {
                    start++;
                }
                else{
                    end--;
                }
            }
        }

        List<List<Integer>> ans_al = new ArrayList<>();

        for(List<Integer> al : ans){
            ans_al.add(al);
        }

        return ans_al;
    }
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        sorted_nums = sorted(nums)

        ans = []

        for i in range(len(sorted_nums)):
            j = i+1
            k = len(sorted_nums) - 1
            
            if i>0 and sorted_nums[i] == sorted_nums[i-1]:
                continue

            while j < k:
                sum1 = sorted_nums[i] + sorted_nums[j] + sorted_nums[k]
                if sum1 == 0:
                    current_ans = [sorted_nums[i], sorted_nums[j], sorted_nums[k]]
                    ans.append(current_ans)
                    j += 1
                    k -= 1
                    while sorted_nums[j] == sorted_nums[j-1] and j<k:
                        j += 1
                elif sum1 < 0:
                    j += 1
                else:
                    k -= 1

        return ans
